storage/inmem/txn.go

```diff
@@ -0,0 +1,410 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package inmem
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/open-policy-agent/opa/storage"
+)
+
+type transaction struct {
+	xid      uint64
+	write    bool
+	log      *snapshot
+	db       *store
+	policies map[string][]byte
+}
+
+type snapshot struct {
+	path   storage.Path
+	remove bool
+	stale  bool
+	value  interface{}
+	next   *snapshot
+	prev   *snapshot
+}
+
+func (txn *transaction) ID() uint64 {
+	return txn.xid
+}
+
+func (txn *transaction) assert(op storage.PatchOp, path storage.Path, value interface{}) error {
+
+	if !txn.write {
```

Comment:
I did not realize right away you don't support upgrading read transactions to read/write transactions. Worth of mentioning in the top transaction comment.